### PR TITLE
Multinomial raise error

### LINKED ISSUE #12260: torch.multinomial without replacement returns repetitive values when all non-zero items are exhausted
## 🐛 Bug

When given a one-hot probability distribution, with replacement=False, torch.multinomial returns 0s for all indices after the first one. While I understand that it is weird to sample > 1 value without replacement if the distribution is one-hot, I think this is less desirable behavior than randomly sampling the remaining values without replacement.

## To Reproduce

``` python
>>> torch.multinomial(torch.FloatTensor([0,1,0,0]), 3, replacement=False)
tensor([1, 0, 0])
```
This seems undesired, because replacement=False implies to me that I should expect my outputs to all be different. My understanding is that this was discussed in a recent issue about the GPU kernel for multinomial not matching the CPU here: https://github.com/pytorch/pytorch/issues/9062

## Expected behavior

I expected not to get the index "0" in the output more than once.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 0.4.0/0.4.1
 - OS (e.g., Linux): 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: 3.6
 - CUDA/cuDNN version: 9
 - GPU models and configuration: K80
 - Any other relevant information: N/A

### PR DESCRIPTION
**Pull request by ailzhang**
Fixes #12260 #2896

```
torch.multinomial(torch.FloatTensor([0, 1, 0, 0]), 3, replacement=False)
```
The old behavior is that we return `0` after we run out of postive categories. Now we raise an error based on discussion in the issue thread. 

- Add testcase for cpu & cuda case, in cuda case `n_samples=1` is a simple special case, so we test against `n_sample=2` instead.

### CODE REVIEW

File: aten/src/TH/generic/THTensorRandom.cpp [hunk 1]
```diff
@@ -310,6 +314,10 @@ void THTensor_(multinomial)(THLongTensor *self, THGenerator *_generator, THTenso
                           THCleanup(THDoubleTensor_free(cum_dist); if (start_dim == 1) THTensor_(squeeze1d)(prob_dist, prob_dist, 0);),
                           2,
                           "invalid multinomial distribution (sum of probabilities <= 0)");
+    THArgCheckWithCleanup((n_categories - n_zeros >= n_sample),
+                          THCleanup(THDoubleTensor_free(cum_dist); if (start_dim == 1) THTensor_(squeeze1d)(prob_dist, prob_dist, 0);),
```
Comment (ezyang) [inline aten/src/TH/generic/THTensorRandom.cpp]: Man, it will be very nice when we can switch this to C++ smart pointer refcounting :)

Comment (ezyang) [reply aten/src/TH/generic/THTensorRandom.cpp]: I'm a little puzzled why there's a call to `squeeze1d` in the cleanup code.

**Review by ezyang — APPROVED** [review]

File: aten/src/TH/generic/THTensorRandom.cpp (hunk 1 above)
Comment (ailzhang) [reply aten/src/TH/generic/THTensorRandom.cpp]: @ezyang  Is it doable in TH/THC? Can you point me to an example code? ;) Thanks!

Comment (ezyang) [reply aten/src/TH/generic/THTensorRandom.cpp]: No, not yet, unless you want to port this entire kernel to ATen :(

Comment (ailzhang) [reply aten/src/TH/generic/THTensorRandom.cpp]: @ezyang We unsqueezed it inplace before. I guess it's mainly for keeping the input not changed. :P 
Btw I created #12499 to port it to Aten, will get to it maybe next week. :)

**Comment by matteorr** [comment]
The documentation at https://pytorch.org/docs/master/torch.html#torch.multinomial does not reflect the change in behavior. The example should be updated to reflect versions 1.x.

**Comment by asanakoy** [comment]
I think it makes sens to print a meaningful error message. I just get `RuntimeError: CUDA error: device-side assert triggered`

` pytorch version  1.1.0`


Code to reproduce:

```
(py36) ~  ipython      
Python 3.6.7 |Anaconda, Inc.| (default, Oct 23 2018, 19:16:44) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch                                                                                                                                                                     

In [2]: weights = torch.zeros(5, 6); weights[:, 3] = 1; weights = weights.cuda()                                                                                                         

In [3]: torch.multinomial(weights, num_samples=1, replacement=False)                                                                                                                     
Out[3]: 
tensor([[3],
        [3],
        [3],
        [3],
        [3]], device='cuda:0')

In [4]: torch.multinomial(weights, num_samples=2, replacement=False)                                                                                                                     
Out[4]: ---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/anaconda2/envs/py36/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/anaconda2/envs/py36/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403 
    404             return _default_pprint(obj, self, cycle)

~/anaconda2/envs/py36/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

~/anaconda2/envs/py36/lib/python3.6/site-packages/torch/tensor.py in __repr__(self)
     69         # characters to replace unicode characters with.
     70         if sys.version_info > (3,):
---> 71             return torch._tensor_str._str(self)
     72         else:
     73             if hasattr(sys.stdout, 'encoding'):

~/anaconda2/envs/py36/lib/python3.6/site-packages/torch/_tensor_str.py in _str(self)
    284                 tensor_str = _tensor_str(self.to_dense(), indent)
    285             else:
--> 286                 tensor_str = _tensor_str(self, indent)
    287 
    288     if self.layout != torch.strided:

~/anaconda2/envs/py36/lib/python3.6/site-packages/torch/_tensor_str.py in _tensor_str(self, indent)
    199     if self.dtype is torch.float16:
    200         self = self.float()
--> 201     formatter = _Formatter(get_summarized_data(self) if summarize else self)
    202     return _tensor_str_with_formatter(self, indent, formatter, summarize)
    203 

~/anaconda2/envs/py36/lib/python3.6/site-packages/torch/_tensor_str.py in __init__(self, tensor)
     81         if not self.floating_dtype:
     82             for value in tensor_view:
---> 83                 value_str = '{}'.format(value)
     84                 self.max_width = max(self.max_width, len(value_str))
     85 

~/anaconda2/envs/py36/lib/python3.6/site-packages/torch/tensor.py in __format__(self, format_spec)
    384     def __format__(self, format_spec):
    385         if self.dim() == 0:
--> 386             return self.item().__format__(format_spec)
    387         return object.__format__(self, format_spec)
    388 

RuntimeError: CUDA error: device-side assert triggered
```

**Comment by soumith** [comment]
@asanakoy due to limitations with CUDA, we cannot print a better error message

**Pull request closed without merging**